x-pack/plugins/global_search/server/services/search_service.ts

```diff
@@ -0,0 +1,119 @@
+import { GlobalSearchConfigType } from '../config';
+import { getContextFactory, GlobalSearchContextFactory } from './context';
+import {
+  GlobalSearchResultProvider,
+  GlobalSearchBatchedResults,
+  GlobalSearchFindOptions,
+} from '../types';
+
+/** @public */
+export interface SearchServiceSetup {
+  registerResultProvider(provider: GlobalSearchResultProvider): void;
+}
+
+/** @public */
+export interface SearchServiceStart {
+  find(
+    term: string,
+    options: GlobalSearchFindOptions,
+    request: KibanaRequest
+  ): Observable<GlobalSearchBatchedResults>;
+}
+
+interface SetupDeps {
+  basePath: IBasePath;
+  licenseChecker: ILicenseChecker;
+  config: GlobalSearchConfigType;
+  maxProviderResults?: number;
+}
+
+const defaultMaxProviderResults = 20;
+
+/** @internal */
+export class SearchService {
+  private readonly providers: GlobalSearchResultProvider[] = [];
+  private basePath?: IBasePath;
+  private config?: GlobalSearchConfigType;
+  private contextFactory?: GlobalSearchContextFactory;
+  private licenseChecker?: ILicenseChecker;
+  private maxProviderResults = defaultMaxProviderResults;
+
+  setup({
+    basePath,
+    config,
+    licenseChecker,
+    maxProviderResults = defaultMaxProviderResults,
+  }: SetupDeps): SearchServiceSetup {
+    this.basePath = basePath;
+    this.config = config;
+    this.licenseChecker = licenseChecker;
+    this.maxProviderResults = maxProviderResults;
+
+    return {
+      registerResultProvider: (provider) => {
+        if (this.providers.map((p) => p.id).includes(provider.id)) {
+          throw new Error(`trying to register duplicate provider: ${provider.id}`);
+        }
+        this.providers.push(provider);
+      },
+    };
+  }
+
+  start(coreStart: CoreStart): SearchServiceStart {
+    this.contextFactory = getContextFactory(coreStart);
+    return {
+      find: (term, options, request) => this.performFind(term, options, request),
+    };
+  }
+
+  private performFind(term: string, options: GlobalSearchFindOptions, request: KibanaRequest) {
+    const licenseState = this.licenseChecker!.getState();
+    if (!licenseState.valid) {
+      return throwError(
+        `GlobalSearch API is disabled because of invalid license state: ${licenseState.message}`
+      );
+    }
+
+    const context = this.contextFactory!(request);
+
+    const timeout$ = timer(this.config!.search_timeout.asMilliseconds()).pipe(map(() => undefined));
+    const aborted$ = options.aborted$ ? merge(options.aborted$, timeout$) : timeout$;
+    const providerOptions = {
+      ...options,
+      preference: options.preference ?? 'default',
+      maxResults: this.maxProviderResults,
+      aborted$,
+    };
+
+    const processResult = (result: GlobalSearchProviderResult) =>
+      processProviderResult(result, this.basePath!);
+
+    const providersResults$ = this.providers.map((provider) =>
+      provider.find(term, providerOptions, context).pipe(
+        takeInArray(this.maxProviderResults),
```

Comment:
Mixed on that. Ideally a user would specify the total maximum number of results he wants, not the number of results per provider (as a consumer don't care and don't have any way to know the number of providers anyway). However as the RFC changed to emit the results per batch, we can't really provide a `maxResults` option anymore. This `maxProviderResults` is more a safe-guard to avoid providers returning hundred of results.